src/bootstrap/src/lib.rs

```diff
@@ -655,27 +655,37 @@ impl Build {
     }
 
     /// Gets the space-separated set of activated features for the standard
-    /// library.
+    /// library. This can be configured with the `std_features` key in config.toml.
     fn std_features(&self, target: TargetSelection) -> String {
-        let mut features = " panic-unwind".to_string();
+        let mut features = if let Some(features) = &self.config.rust_std_features {
+            features.iter().map(|s| s.as_ref()).collect::<Vec<&str>>()
+        } else {
+            vec![]
+        };
+
+        if !features.contains(&"panic-unwind") {
+            features.push("panic-unwind");
+        }
 
         match self.config.llvm_libunwind(target) {
-            LlvmLibunwind::InTree => features.push_str(" llvm-libunwind"),
-            LlvmLibunwind::System => features.push_str(" system-llvm-libunwind"),
+            LlvmLibunwind::InTree => features.push("llvm-libunwind"),
+            LlvmLibunwind::System => features.push("system-llvm-libunwind"),
             LlvmLibunwind::No => {}
         }
         if self.config.backtrace {
-            features.push_str(" backtrace");
+            features.push("backtrace");
         }
         if self.config.profiler_enabled(target) {
-            features.push_str(" profiler");
+            features.push("profiler");
         }
         // Generate memcpy, etc.  FIXME: Remove this once compiler-builtins
         // automatically detects this target.
         if target.contains("zkvm") {
-            features.push_str(" compiler-builtins-mem");
+            features.push("compiler-builtins-mem");
         }
-        features
+
+        // remove duplicates
+        features.into_iter().collect::<HashSet<_>>().into_iter().collect::<Vec<_>>().join(" ")
```

Comment:
Please go through `BtreeSet` or use explicit sort+dedup instead. `HashSet` will produce a non-deterministic order of the feature flags.